src/Orleans/Threading/GrainCancellationTokenSource.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Orleans.Threading
+{
+    /// <summary>
+    /// Distributed version of the CancellationTokenSource
+    /// </summary>
+    public class GrainCancellationTokenSource : IDisposable
+    {
+        private readonly CancellationTokenSource _cancellationTokenSource;
+
+        private readonly GrainCancellationToken _grainCancellationToken;
+
+        /// <summary>
+        /// Initializes the <see cref="T:Orleans.Threading.GrainCancellationTokenSource"/>.
+        /// </summary>
+        public GrainCancellationTokenSource() : this(Guid.NewGuid(), false)
+        {
+        }
+
+        internal GrainCancellationTokenSource(Guid id, bool cancelled)
```

Comment:
nit picking: the .NET framework uses double L for spelling "cancellation", but it always uses a single L for spelling "Canceled". I think we should imitate that. See [here](https://msdn.microsoft.com/en-us/library/system.threading.cancellationtoken(v=vs.110).aspx) and [here](https://msdn.microsoft.com/en-us/library/system.threading.tasks.taskstatus(v=vs.110).aspx) for examples